dataflowengineoss/src/main/scala/io/joern/dataflowengineoss/queryengine/SourcesToStartingPoints.scala

```diff
@@ -9,48 +9,132 @@ import io.shiftleft.semanticcpg.language.operatorextension.allAssignmentTypes
 import io.shiftleft.semanticcpg.utils.MemberAccess.isFieldAccess
 import org.slf4j.LoggerFactory
 
-import java.util.concurrent.{ForkJoinPool, ForkJoinTask, RecursiveTask, RejectedExecutionException}
+import java.util.concurrent.*
+import scala.collection.mutable.ListBuffer
 import scala.util.{Failure, Success, Try}
 
 case class StartingPointWithSource(startingPoint: CfgNode, source: StoredNode)
-
+case class UsageInput(src: StoredNode, typeDecl: TypeDecl, astNode: AstNode)
+case class ResultSummary(result: List[StartingPointWithSource], methodTasks: List[UsageInput])
 object SourcesToStartingPoints {
 
   private val log = LoggerFactory.getLogger(SourcesToStartingPoints.getClass)
 
   def sourceTravsToStartingPoints[NodeType](sourceTravs: IterableOnce[NodeType]*): List[StartingPointWithSource] = {
-    val fjp = ForkJoinPool.commonPool()
+    val executorService = Executors.newWorkStealingPool()
+    val completionService =
+      new ExecutorCompletionService[Unit](executorService)
     try {
-      fjp.invoke(new SourceTravsToStartingPointsTask(sourceTravs: _*)).distinct
+      val sources: List[StoredNode] = sourceTravs
+        .flatMap(_.iterator.toList)
+        .collect { case n: StoredNode => n }
+        .dedup
+        .toList
+        .sortBy(_.id)
+      val firstResultAggregator       = SourceStartingPointResultAggregator(sources.size)
+      val firstResultAggregatorThread = new Thread(firstResultAggregator)
+      firstResultAggregatorThread.setName("Starting points first result aggregator")
+      firstResultAggregatorThread.start()
+      sources.foreach(src =>
+        completionService.submit(new SourceToStartingPoints(src, firstResultAggregator.resultQueue))
+      )
+      firstResultAggregatorThread.join()
+      val secondResults = sources.headOption
+        .map(src => {
+          val cpg                          = Cpg(src.graph())
```

Comment:
I need to get the one source node from which I can create CPG wrapper as explained in the above comment. It would be great if you could help me understand how can I move these two statements in the parent scope safely with the assumption that `sources` list could be an empty list.